apps/fhir/server/authentication.py

```diff
@@ -1,38 +1,125 @@
 import requests
-import logging
+from django.conf import settings
 from rest_framework import exceptions
 from ..bluebutton.exceptions import UpstreamServerException
 from ..bluebutton.utils import (FhirServerAuth,
                                 get_resourcerouter)
+from .loggers import log_match_fhir_id
 
-logger = logging.getLogger('hhs_server.%s' % __name__)
 
+def search_fhir_id_by_identifier_mbi_hash(mbi_hash):
+    """
+        Search the backend FHIR server's patient resource
+        using the mbi_hash identifier.
+    """
+    search_identifier = settings.FHIR_SEARCH_PARAM_IDENTIFIER_MBI_HASH \
+        + "%7C" + mbi_hash
 
-def match_hicn_hash(hicn_hash):
-    auth_state = FhirServerAuth(None)
-    certs = (auth_state['cert_file'], auth_state['key_file'])
-    # URL for patient ID.
-    url = get_resourcerouter().fhir_url + \
-        "Patient/?identifier=http%3A%2F%2Fbluebutton.cms.hhs.gov%2Fidentifier%23hicnHash%7C" + \
-        hicn_hash + \
-        "&_format=json"
+    return search_fhir_id_by_identifier(search_identifier)
+
+
+def search_fhir_id_by_identifier_hicn_hash(hicn_hash):
+    """
+        Search the backend FHIR server's patient resource
+        using the hicn_hash identifier.
+    """
+    search_identifier = settings.FHIR_SEARCH_PARAM_IDENTIFIER_HICN_HASH \
+        + "%7C" + hicn_hash
+
+    return search_fhir_id_by_identifier(search_identifier)
+
+
+def search_fhir_id_by_identifier(search_identifier):
+    """
+        Search the backend FHIR server's patient resource
+        using the specified identifier.
+
+        Return:  fhir_id = matched ID (or None).
+                 err_mesg = None or related error message.
+    """
+    # Get certs from FHIR server settings
+    auth_settings = FhirServerAuth(None)
+    certs = (auth_settings['cert_file'], auth_settings['key_file'])
+
+    # Build URL with patient ID search by identifier.
+    url = get_resourcerouter().fhir_url \
+        + "Patient/?identifier=" + search_identifier \
+        + "&_format=" + settings.FHIR_PARAM_FORMAT
+
+    # Get FHIR service backend response.
+    #   TODO: Should work with verify=True
     response = requests.get(url, cert=certs, verify=False)
     response.raise_for_status()
     backend_data = response.json()
 
-    if backend_data.get('total', 0) > 1:
-        # Don't return a 404 because retrying later will not fix this.
-        raise UpstreamServerException("Duplicate beneficiaries found")
+    # Parse and validate backend_data response.
+    if (
+        'entry' in backend_data
+            and backend_data.get('total', 0) == 1
+            and len(backend_data.get('entry', '')) == 1
```

Comment:
@karlmdavis This is an interesting case.  Below is an example JSON for `entry` (with other stuff stripped out) for reference.  The response example you gave would match the conditions for the code to run at line #62.  However, it would get a KeyError when trying to access: `backend_data['entry'][0]['resource']['id']`

```json
            "entry":[
                {
                    "resource":{
                        "resourceType":"Patient",
                        "id":"-20000000002346",
                        "extension":[
                            {
                                }
                        ],
```

I'll add some more validation logic for the expected result. I'm thinking of the following:
* Check that `backend_data['entry']` is a List type with only a single element.
* Check that `backend_data['entry'][0]['resource']` is available.
* Check that `backend_data['entry'][0]['resource']['type'] == 'Patient'.
* Check that `backend_data['entry'][0]['resource']['id']` is an expected FHIR_ID that is an integer value.

Reply:
@karlmdavis I refactored the code to add some more conditional and type checks on the expected (happy path) response from the backend.  If the structure of the response doesn't match what's expected, it will fall through to the "Unexpected result ..." exception.

Reply:
That's likely not very Pyhtonic, but given how critical that one if clause is, also likely a very good idea.